### PR TITLE
Make TimeArrays as PyTrees

### PR DESCRIPTION
**Pull request by gautierronan**
Makes all TimeArrays subclasses of eqx.Module such that they are Pytrees and that the JIT is well-handled.

With this syntax, `dq.totime` should receive a callable of signature `(t: Scalar, args: tuple) -> Array` instead of `(t: Scalar) -> Array` as before. The `args` can then be passed into `dq.totime` as an optional argument, and the jit doesn't run again.

Note we probably don't need `eqx.Module` and could work around it, but it makes everything very neat IMO.

To test this:
```python
import jax.numpy as jnp
import time
from jax import jit
import dynamiqs as dq


def time_once(f, *args, **kwargs):
    start = time.time()
    f(*args, **kwargs).block_until_ready()
    print(f'Elapsed: {(time.time() - start) * 1000:.3f} ms')


@jit
def loss(x):
    return jnp.linalg.norm(2 * x(1.0) - x(0.0))


# ==============================
# test constant array
# ==============================

N = 100
x0 = jnp.arange(N).repeat(N).reshape(N, N).astype(complex)
x1 = (N - jnp.arange(N)).repeat(N).reshape(N, N).astype(complex)
array0 = dq.totime(x0)
array1 = dq.totime(x1)

print("=== constant array")
array = array0
time_once(loss, array)  # jit runs

print("=== constant array (different array)")
array = array0 + array1
time_once(loss, array)  # jit does not rerun

# ==============================
# test callable array
# ==============================

fn = lambda t, args: args * t * x0
array2 = dq.totime(fn, args=2.0)
array3 = dq.totime(fn, args=3.0)

print("\n=== callable array")
array = array0 + array1 + array2
time_once(loss, array)  # jit runs

print("=== callable array (same function with different parameter)")
array = array0 + array1 + array3
time_once(loss, array)  # jit does not rerun

print("=== callable array (different function)")
array = array0 + array1 + array2 + array3
time_once(loss, array)  # jit runs

# ==============================
# test PWC array
# ==============================

times = jnp.linspace(0.0, 1.0, 11)
values0 = jnp.arange(10).astype(complex)
values1 = (10 - jnp.arange(10)).astype(complex)
array4 = dq.totime((times, values0, x0))
array5 = dq.totime((times, values1, x0))
array6 = dq.totime((times, values0, x1))

print("\n=== PWC array")
array = array4
time_once(loss, array)  # jit runs

print("=== PWC array (different values)")
array = array5
time_once(loss, array)  # jit does not rerun

print("=== PWC array (different array)")
array = array6
time_once(loss, array)  # jit does not rerun

print("=== PWC array (sum of two PWC arrays)")
array = array4 + array5
time_once(loss, array)  # jit runs

# ==============================
# test modulated array
# ==============================

fn = lambda t, args: args * jnp.array(t)
array7 = dq.totime((fn, x0), args=1.0)
array8 = dq.totime((fn, x0), args=2.0)

print("\n=== modulated array")
array = array7
time_once(loss, array)  # jit runs

print("=== modulated array (same function with different parameter)")
array = array8
time_once(loss, array)  # jit does not rerun

print("=== modulated array (different function)")
array = array7 + array8
time_once(loss, array)  # jit runs
```
which prints
```markdown
=== constant array
Elapsed: 24.352 ms
=== constant array (different array)
Elapsed: 0.045 ms

=== callable array
Elapsed: 26.352 ms
=== callable array (same function with different parameter)
Elapsed: 0.066 ms
=== callable array (different function)
Elapsed: 28.494 ms

=== PWC array
Elapsed: 72.702 ms
=== PWC array (different values)
Elapsed: 0.074 ms
=== PWC array (different array)
Elapsed: 0.050 ms
=== PWC array (sum of two PWC arrays)
Elapsed: 123.167 ms

=== modulated array
Elapsed: 29.345 ms
=== modulated array (same function with different parameter)
Elapsed: 0.072 ms
=== modulated array (different function)
Elapsed: 51.495 ms
```

In a subsequent PR:

- [ ] Move these tests into our test set.

### CODE REVIEW

File: dynamiqs/time_array.py [hunk 1]
```diff
@@ -156,12 +163,20 @@ def _factory_modulated(
...
 
-    factors = [_ModulatedFactor(f, f0)]
+    factors = [_ModulatedFactor(Partial(f, args=args), f0)]
     arrays = array[None, ...]  # (1, n, n)
-    return ModulatedTimeArray(factors, arrays)
+    return ModulatedTimeArray(factors, arrays, static=jnp.zeros_like(array))
 
 
-class TimeArray:
+class SumCallable(eqx.Module):
```
Comment (abocquet) [inline dynamiqs/time_array.py:171]: Looks like this could be a `CallableTimeArray`

**Review by abocquet — CHANGES_REQUESTED** [review]
Nice job !
Few remarks:
- We could add (in a future PR) support for lambdas without the `args` argument
- Does it stop working if TimeArray is no longer an equinox module ? (I'm still trying to figure out what it does)

**Review by abocquet — COMMENTED** [review]
.

**Comment by gautierronan** [comment]
>   * We could add (in a future PR) support for lambdas without the `args` argument

We could, but I'm not sure we want to? Maybe having a strict function signature is better?

>   * Does it stop working if TimeArray is no longer an equinox module ? (I'm still trying to figure out what it does)

Yes, I get the following error:
```python
---------------------------------------------------------------------------
TypeError                                 Traceback (most recent call last)
[/Users/rgautier/Code/dynamiqs/notebooks/_implem_pytree_timearray.py](https://file+.vscode-resource.vscode-cdn.net/Users/rgautier/Code/dynamiqs/notebooks/_implem_pytree_timearray.py) in line 24
     [22](file:///Users/rgautier/Code/dynamiqs/notebooks/_implem_pytree_timearray.py?line=21) print("=== constant array")
     [23](file:///Users/rgautier/Code/dynamiqs/notebooks/_implem_pytree_timearray.py?line=22) array = array0
---> [24](file:///Users/rgautier/Code/dynamiqs/notebooks/_implem_pytree_timearray.py?line=23) time_once(loss, array)  # jit runs
     [25](file:///Users/rgautier/Code/dynamiqs/notebooks/_implem_pytree_timearray.py?line=24) print("=== constant array (different array)")
     [26](file:///Users/rgautier/Code/dynamiqs/notebooks/_implem_pytree_timearray.py?line=25) array = array0 + array1

[/Users/rgautier/Code/dynamiqs/notebooks/_implem_pytree_timearray.py](https://file+.vscode-resource.vscode-cdn.net/Users/rgautier/Code/dynamiqs/notebooks/_implem_pytree_timearray.py) in line 7, in time_once(f, *args, **kwargs)
      [5](file:///Users/rgautier/Code/dynamiqs/notebooks/_implem_pytree_timearray.py?line=4) def time_once(f, *args, **kwargs):
      [6](file:///Users/rgautier/Code/dynamiqs/notebooks/_implem_pytree_timearray.py?line=5)     start = time.time()
----> [7](file:///Users/rgautier/Code/dynamiqs/notebooks/_implem_pytree_timearray.py?line=6)     f(*args, **kwargs).block_until_ready()
      [8](file:///Users/rgautier/Code/dynamiqs/notebooks/_implem_pytree_timearray.py?line=7)     print(f'Elapsed: {(time.time() - start) * 1000:.3f} ms')

    [... skipping hidden 12 frame]

[/Users/rgautier/Code/dynamiqs/notebooks/_implem_pytree_timearray.py](https://file+.vscode-resource.vscode-cdn.net/Users/rgautier/Code/dynamiqs/notebooks/_implem_pytree_timearray.py) in line 12, in loss(x)
     [10](file:///Users/rgautier/Code/dynamiqs/notebooks/_implem_pytree_timearray.py?line=9) @jit
     [11](file:///Users/rgautier/Code/dynamiqs/notebooks/_implem_pytree_timearray.py?line=10) def loss(x):
---> [12](file:///Users/rgautier/Code/dynamiqs/notebooks/_implem_pytree_timearray.py?line=11)     return jnp.linalg.norm(2 * x(1.0) - x(0.0))

TypeError: 'DynamicJaxprTracer' object is not callable
```

File: dynamiqs/time_array.py (hunk 1 above)
Comment (gautierronan) [reply dynamiqs/time_array.py:171]: I don't think so. It's really just a trick to be able to sum callables while maintaining the jit. It was actually proposed by @patrick-kidger.

File: dynamiqs/time_array.py [hunk 2]
```diff
@@ -22,22 +24,26 @@ def totime(
         | tuple[callable[[float], Array], ArrayLike]
     ),
     *,
+    args: tuple = (),
```
Comment (pierreguilmin) [inline dynamiqs/time_array.py:27]: You need to edit the typing `callable[[float], Array]` of `totime` signature.

File: dynamiqs/time_array.py [hunk 3]
```diff
@@ -22,22 +24,26 @@ def totime(
...
     dtype: jnp.dtype | None = None,
 ) -> TimeArray:
     dtype = dtype or get_cdtype(dtype)  # assume complex by default
 
+    # already a time array
+    if isinstance(x, TimeArray):
```
Comment (pierreguilmin) [inline dynamiqs/time_array.py:33]: You need to include this in allowed types for `x` in `totime` signature..

File: dynamiqs/time_array.py [hunk 4]
```diff
@@ -73,7 +79,7 @@ def _factory_callable(
             ' conversion at each solver time step.'
         )
 
-    return CallableTimeArray(x, f0)
+    return CallableTimeArray(Partial(f, args=args), f0)
```
Comment (pierreguilmin) [inline dynamiqs/time_array.py:82]: Maybe a line comment about this magic would be great (in the line on "Partial = jax way to do callable pytrees").

File: dynamiqs/time_array.py [hunk 5]
```diff
@@ -376,26 +385,24 @@ def conj(self) -> _Factor:
...
-        else:
-            # find the index $k$ such that $t \in [t_k, t_{k+1})$
-            idx = jnp.searchsorted(self.times, t, side='right') - 1
-            return self.values[..., idx]  # (...)
+        idx = jnp.searchsorted(self.times, t, side='right') - 1
+        return lax.select(
+            jnp.logical_or(t < self.times[0], t >= self.times[-1]),
+            jnp.zeros_like(self.values[..., 0]),
+            self.values[..., idx],
+        )
```
Comment (pierreguilmin) [inline dynamiqs/time_array.py:393]: See https://github.com/dynamiqs/dynamiqs/pull/410#discussion_r1447122100.

File: dynamiqs/time_array.py [hunk 6]
```diff
@@ -416,22 +423,25 @@ def reshape(self, *args: int) -> _Factor:
...
-        self, factors: list[_Factor], arrays: Array, static: Array | None = None
-    ):
-        # factors must be non-empty
-        self.factors = factors  # list of length (nf)
-        self.arrays = arrays  # (nf, n, n)
-        self.n = arrays.shape[-1]
-        self.static = jnp.zeros_like(self.arrays[0]) if static is None else static
+    factors: list[_Factor]
+    arrays: Array
+    static: Array
```
Comment (pierreguilmin) [inline dynamiqs/time_array.py:428]: Can we keep the shape comment, and the `# factors must be non-empty`.

File: dynamiqs/time_array.py [hunk 7]
```diff
@@ -416,22 +423,25 @@ def reshape(self, *args: int) -> _Factor:
...
-        return tuple(self.factors[0].shape, self.n, self.n)  # (..., n, n)
+        n = self.arrays.shape[-1]
+        return tuple(self.factors[0].shape, n, n)  # (..., n, n)
+
+    @property
+    def mT(self) -> TimeArray:
+        factors = [x.conj() for x in self.factors]
+        arrays = self.arrays.conj().mT
+        static = self.static.conj().mT
+        return FactorTimeArray(factors, arrays, static=static)
```
Comment (pierreguilmin) [inline dynamiqs/time_array.py:444]: mT should not conj.

File: dynamiqs/time_array.py [hunk 8]
```diff
@@ -471,12 +481,7 @@ def __add__(self, y: ArrayLike | TimeArray) -> TimeArray:
 
 class PWCTimeArray(FactorTimeArray):
     # Arbitrary sum of arrays with PWC factors.
-
-    def __init__(self, *args, **kwargs):
-        super().__init__(*args, **kwargs)
-
-        # merge all times
-        self.times = jnp.unique(jnp.concatenate([x.times for x in self.factors]))
+    pass
```
Comment (pierreguilmin) [inline dynamiqs/time_array.py:484]: We need `times` for `PWCTimeArray`.

**Review by pierreguilmin — CHANGES_REQUESTED** [review]
Very cool 🥳 Yes equinox is just syntactic sugar, but why not, it looks nice.

**Review by pierreguilmin — DISMISSED** [review]

**Review by abocquet — APPROVED** [review]
LGTM 👍

**Pull request merged**